File: thonny-projects/Pico-DHT22-MQTT/network_utils.py
"""
网络工具模块
提供 WiFi 连接和 NTP 时间同步功能
"""

import time
import network, ubinascii


class WiFiManager:
    """WiFi 连接管理器"""
    
    def __init__(self, ssid, password):
        """
        初始化 WiFi 管理器
        
        Args:
            ssid: WiFi 网络名称
            password: WiFi 密码
        """
        self.ssid = ssid
        self.password = password
        self.wlan = None
    
    def connect(self, timeout=30, watchdog_feed_callback=None):
        """
        连接到 WiFi 网络
        
        Args:
            timeout: 连接超时时间（秒），默认 30 秒
            watchdog_feed_callback: 看门狗喂狗回调函数（可选）
            
        Returns:
            bool:  连接成功返回 True，失败返回 False
        """
        try:
            self.wlan = network.WLAN(network.STA_IF)
            self.wlan.config(pm=network.WLAN.PM_NONE)  # 禁用电源管理
            self.wlan.active(True)
            
            # 如果已经连接，直接返回
            if self.wlan.isconnected():
                print(f"已连接到 WiFi: {self.wlan.ifconfig()[0]}")
                return True
            
            # 开始连接
            print(f"正在连接到 WiFi: {self.ssid}")
            self.wlan.connect(self.ssid, self.password)
            
            # 等待连接（每秒喂狗）
            start_time = time.time()
            while not self.wlan.isconnected():
                if time.time() - start_time > timeout:
                    print(f"WiFi 连接超时（{timeout}秒）")
                    return False
                
                # 喂狗（如果提供了回调函数）
                if watchdog_feed_callback:
                    watchdog_feed_callback()
                
                print("等待连接...")
                time.sleep(1)
            
            # 连接成功
            ip_address = self.wlan.ifconfig()[0]
            print(f"WiFi 连接成功！IP 地址: {ip_address}")
            return True
            
        except Exception as e:
            print(f"WiFi 连接失败: {e}")
            return False
    
    def disconnect(self):
        """断开 WiFi 连接"""
        if self.wlan and self.wlan.isconnected():
            self.wlan.disconnect()
            self.wlan.active(False)
            print("WiFi 已断开")
    
    def is_connected(self):
        """
        检查 WiFi 是否已连接
        
        Returns:
            bool: 已连接返回 True，否则返回 False
        """
        return self.wlan and self.wlan.isconnected()
    
    def get_ip_address(self):
        """
        获取 IP 地址
        
        Returns:
            str: IP 地址，未连接返回 None
        """
        if self.is_connected():
            return self.wlan.ifconfig()[0]
        return None
    
    def get_network_info(self):
        """
        获取网络信息
        
        Returns:
            dict: 包含 IP、子网掩码、网关、DNS 的字典
        """
        if self.is_connected():
            ifconfig = self.wlan.ifconfig()
            return {
                'ip': ifconfig[0],
                'subnet':  ifconfig[1],
                'gateway': ifconfig[2],
                'dns': ifconfig[3],
                'rssi': self.wlan.status('rssi'),
                'mac': ubinascii.hexlify(self.wlan.config('mac'), ':').decode()
            }
        return None
    
    def cleanup(self):
        """
        清理资源并释放内存
        """
        try:
            if self.wlan:
                if self.wlan.isconnected():
                    self.wlan.disconnect()
                self.wlan.active(False)
                self.wlan = None
                print("WiFi 资源已清理")
        except Exception as e:
            print(f"WiFi 清理异常: {e}")


class NTPTimeSync:
    """NTP 时间同步管理器"""
    
    # 常用 NTP 服务器列表
    NTP_SERVERS = [
        "ntp.aliyun.com",       # 阿里云
        "ntp.ntsc.ac.cn",       # 中国国家授时中心
        "ntp1.aliyun.com",      # 阿里云备用
        "pool.ntp.org",         # 国际 NTP 池
    ]
    
    def __init__(self, timezone_offset=8):
        """
        初始化 NTP 时间同步器
        
        Args:
            timezone_offset: 时区偏移量（小时），默认 8（北京时间 UTC+8）
        """
        self.timezone_offset = timezone_offset
    
    def sync(self, ntp_server=None, retry_count=3):
        """
        从 NTP 服务器同步时间
        
        Args:
            ntp_server:  NTP 服务器地址，默认使用服务器列表
            retry_count: 重试次数，默认 3 次
            
        Returns: 
            bool: 同步成功返回 True，失败返回 False
        """
        try:
            import ntptime
            from machine import RTC
            
            # 确定要使用的 NTP 服务器
            servers = [ntp_server] if ntp_server else self.NTP_SERVERS
            
            # 尝试每个服务器
            for server in servers:
                for attempt in range(retry_count):
                    try:
                        print(f"正在从 {server} 同步时间...  (尝试 {attempt + 1}/{retry_count})")
                        ntptime.host = server
                        ntptime.settime()
                        
                        # 同步成功，调整时区
                        self._adjust_timezone()
                        
                        current_time = self. get_iso8601_time()
                        print(f"时间同步成功: {current_time}")
                        return True
                        
                    except Exception as e: 
                        print(f"同步失败: {e}")
                        if attempt < retry_count - 1:
                            time.sleep(2)  # 重试前等待
                        continue
            
            print("所有 NTP 服务器同步失败")
            return False
            
        except ImportError:
            print("错误:  ntptime 模块不可用")
            return False
        except Exception as e:
            print(f"时间同步异常: {e}")
            return False
    
    def _adjust_timezone(self):
        """调整系统时间到指定时区"""
        try: 
            from machine import RTC
            
            rtc = RTC()
            utc_timestamp = time.mktime(time.localtime())
            local_timestamp = utc_timestamp + self.timezone_offset * 3600
            local_time = time.localtime(local_timestamp)
            
            rtc.datetime((
                local_time[0],  # 年
                local_time[1],  # 月
                local_time[2],  # 日
                local_time[6],  # 星期
                local_time[3],  # 时
                local_time[4],  # 分
                local_time[5],  # 秒
                0               # 微秒
            ))
            
        except Exception as e:
            print(f"时区调整失败: {e}")
    
    def get_iso8601_time_local(self):
        """获取本地时间的 ISO 8601 格式（带时区偏移）"""
        t = time.localtime()
        # 计算时区偏移的小时和分钟
        offset_hours = self.timezone_offset // 3600
        offset_minutes = abs(self.timezone_offset % 3600) // 60
        offset_sign = '+' if offset_hours >= 0 else '-'
        
        return "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}{}{:02d}:{:02d}". format(
            t[0], t[1], t[2], t[3], t[4], t[5],
            offset_sign, abs(offset_hours), offset_minutes
        )
    
    def get_iso8601_time_with_timezone(self, timezone_offset = 8 * 3600):
        """
        获取带时区偏移的 ISO 8601 格式时间
        : param timezone_offset: 时区偏移（秒），例如 UTC+8 = 8*3600
        : return: 格式如 2026-01-04T09:49:41+08:00
        """
        # 获取当前本地时间
        t = time.localtime()
        
        # 计算时区偏移的小时和分钟
        offset_hours = timezone_offset // 3600
        offset_minutes = abs(timezone_offset % 3600) // 60
        offset_sign = '+' if offset_hours >= 0 else '-'
        
        # 格式化为 ISO 8601 带时区偏移
        return "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}{}{:02d}:{:02d}".format(
            t[0], t[1], t[2], t[3], t[4], t[5],
            offset_sign, abs(offset_hours), offset_minutes
        )
    
    @staticmethod
    def get_iso8601_time():
        """
        获取 ISO 8601 格式的当前时间字符串
        
        Returns:
            str: ISO 8601 格式时间 (YYYY-MM-DDTHH:MM:SS)
        """
        current_time = time.localtime()
        
        iso8601_time = "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}".format(
            current_time[0],  # 年
            current_time[1],  # 月
            current_time[2],  # 日
            current_time[3],  # 时
            current_time[4],  # 分
            current_time[5]   # 秒
        )
        
        return iso8601_time
    
    @staticmethod
    def get_timestamp():
        """
        获取当前时间戳
        
        Returns:
            int:  Unix 时间戳
        """
        return time.mktime(time.localtime())
    
    @staticmethod
    def get_iso8601_time_utc(timezone_offset = 8 * 3600):
        """
        获取 UTC 时间的 ISO 8601 格式
        :param timezone_offset: 当前系统时间的时区偏移（秒）
        """
        # 获取当前系统时间（本地时间）
        local_time = time.time()
    
        # 减去时区偏移，得到 UTC 时间
        t = time.gmtime(local_time - timezone_offset)  # 使用 gmtime() 获取 UTC 时间
        return "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}Z".format(
            t[0], t[1], t[2], t[3], t[4], t[5]
        )
    
    @staticmethod
    def format_time(format_str="%Y-%m-%d %H:%M:%S"):
        """
        格式化当前时间
        
        Args:
            format_str: 时间格式字符串
            
        Returns:
            str: 格式化的时间字符串
        """
        current_time = time.localtime()
        
        # MicroPython 不支持 strftime，手动格式化
        formatted_time = format_str.replace("%Y", "{:04d}".format(current_time[0]))
        formatted_time = formatted_time.replace("%m", "{:02d}".format(current_time[1]))
        formatted_time = formatted_time.replace("%d", "{:02d}".format(current_time[2]))
        formatted_time = formatted_time.replace("%H", "{:02d}".format(current_time[3]))
        formatted_time = formatted_time.replace("%M", "{:02d}".format(current_time[4]))
        formatted_time = formatted_time.replace("%S", "{:02d}".format(current_time[5]))
        
        return formatted_time


# ==================== 便捷函数 ====================

def quick_connect_wifi(ssid, password, timeout=30):
    """
    快速连接 WiFi（便捷函数）
    
    Args:
        ssid:  WiFi 网络名称
        password: WiFi 密码
        timeout: 连接超时时间（秒）
        
    Returns:
        WiFiManager: WiFi 管理器实例，连接失败返回 None
    """
    wifi = WiFiManager(ssid, password)
    if wifi.connect(timeout):
        return wifi
    return None


def quick_sync_time(timezone_offset=8, ntp_server=None):
    """
    快速同步时间（便捷函数）
    
    Args: 
        timezone_offset: 时区偏移量（小时）
        ntp_server:  NTP 服务器地址
        
    Returns:
        bool: 同步成功返回 True，失败返回 False
    """
    ntp = NTPTimeSync(timezone_offset)
    return ntp.sync(ntp_server)

File: arduino-projects/esp32_dht22_mqtt/NetworkUtils.cpp
/**
 * 网络工具模块 - ESP32-C3版本实现
 * 提供WiFi连接和NTP时间同步功能
 * 作者: Arduino版本转换
 * 日期: 2026年1月
 */

#include "NetworkUtils.h"
#include <time.h>
#include <sys/time.h>

// ==================== WiFiManager 实现 ====================

// 构造函数
WiFiManager::WiFiManager(const String& ssid, const String& password)
    : ssid(ssid), password(password), connected(false), logCallback(nullptr) {
}

// 析构函数
WiFiManager::~WiFiManager() {
    cleanup();
}

// 设置日志回调函数
void WiFiManager::setLogCallback(LogCallback callback) {
    logCallback = callback;
}

// 内部日志方法
void WiFiManager::log(const String& message) {
    if (logCallback) {
        logCallback(message);
    } else {
        Serial.println(message);
    }
}

// 获取WiFi状态字符串
String WiFiManager::getStatusString(int status) {
    switch (status) {
        case WL_IDLE_STATUS:
            return "空闲状态";
        case WL_NO_SSID_AVAIL:
            return "无可用网络";
        case WL_SCAN_COMPLETED:
            return "扫描完成";
        case WL_CONNECTED:
            return "已连接";
        case WL_CONNECT_FAILED:
            return "连接失败";
        case WL_CONNECTION_LOST:
            return "连接丢失";
        case WL_DISCONNECTED:
            return "已断开";
        case WL_NO_SHIELD:
            return "无WiFi模块";
        default:
            return "未知状态";
    }
}

// 连接到WiFi网络
bool WiFiManager::connect(int timeout, int maxRetries, int retryDelay, WatchdogCallback watchdogCallback) {
    try {
        // 配置WiFi模式
        WiFi.mode(WIFI_STA);
        WiFi.setSleep(false);  // 禁用WiFi省电模式

        // 启用Wi-Fi自动重连机制
        // WiFi.setAutoConnect(true);
        // 启用DHCP以确保IP地址动态分配
        WiFi.config(INADDR_NONE, INADDR_NONE, INADDR_NONE, INADDR_NONE);
        
        // 如果已经连接，直接返回
        if (WiFi.status() == WL_CONNECTED) {
            connected = true;
            log("已连接到 WiFi: " + WiFi.localIP().toString());
            return true;
        }
        
        log("开始WiFi连接，最大重试次数: " + String(maxRetries));
        
        // 重试连接循环
        for (int attempt = 1; attempt <= maxRetries; attempt++) {
            log("=== 连接尝试 " + String(attempt) + "/" + String(maxRetries) + " ===");
            
            // 清理之前的连接状态
            if (attempt > 1) {
                log("清理WiFi状态...");
                WiFi.disconnect(true);
                delay(1000);
                WiFi.mode(WIFI_STA);
                delay(500);
            }
            
            // 先扫描并验证目标SSID是否存在（第一次尝试或网络扫描失败时）
            if (attempt == 1 || attempt % 2 == 0) {
                log("扫描网络并验证SSID...");
                if (!this->scanAndVerifySSID()) {
                    if (attempt < maxRetries) {
                        log("网络扫描失败，等待 " + String(retryDelay/1000) + " 秒后重试...");
                        delay(retryDelay);
                        continue;
                    } else {
                        log("所有重试均失败：未找到目标网络");
                        connected = false;
                        return false;
                    }
                }
            }
            
            // 开始连接
            log("正在连接到 WiFi: " + ssid + " (尝试 " + String(attempt) + ")");
            WiFi.begin(ssid.c_str(), password.c_str());
            
            // 等待连接（每秒检查状态并喂狗）
            unsigned long startTime = millis();
            bool connectionSuccess = false;
            
            while (WiFi.status() != WL_CONNECTED) {
                if (millis() - startTime > timeout * 1000UL) {
                    log("连接超时（" + String(timeout) + "秒）");
                    break;
                }
                
                // 喂狗（如果提供了回调函数）
                if (watchdogCallback) {
                    watchdogCallback();
                }
                
                int status = WiFi.status();
                log("等待连接... 状态: " + String(status) + " (" + getStatusString(status) + ")");
                
                // 检查特定的错误状态
                if (status == WL_CONNECT_FAILED) {
                    log("连接失败，可能是密码错误");
                    break;
                } else if (status == WL_NO_SSID_AVAIL) {
                    log("网络不可用");
                    break;
                }
                
                delay(1000);
            }
            
            // 检查连接结果
            if (WiFi.status() == WL_CONNECTED) {
                connected = true;
                String ipAddress = WiFi.localIP().toString();
                int rssi = WiFi.RSSI();
                log("✓ WiFi 连接成功！");
                log("  IP 地址: " + ipAddress);
                log("  信号强度: " + String(rssi) + " dBm");
                log("  尝试次数: " + String(attempt));
                return true;
            } else {
                // 连接失败
                int status = WiFi.status();
                log("❌ 连接失败，状态: " + String(status) + " (" + getStatusString(status) + ")");
                
                if (attempt < maxRetries) {
                    log("等待 " + String(retryDelay/1000) + " 秒后重试...");
                    delay(retryDelay);
                } else {
                    log("所有重试均失败，放弃连接");
                }
            }
        }
        
        // 所有重试都失败了
        connected = false;
        log("WiFi 连接失败：已达到最大重试次数");
        return false;
        
    } catch (...) {
        log("WiFi 连接失败: 未知异常");
        connected = false;
        return false;
    }
}

// 断开WiFi连接
void WiFiManager::disconnect() {
    if (WiFi.status() == WL_CONNECTED) {
        WiFi.disconnect(true);
        WiFi.mode(WIFI_OFF);
        connected = false;
        log("WiFi 已断开");
    }
}

// 检查WiFi是否已连接
bool WiFiManager::isConnected() {
    bool wifiConnected = (WiFi.status() == WL_CONNECTED);
    connected = wifiConnected;
    return connected;
}

// 获取IP地址
String WiFiManager::getIpAddress() {
    if (isConnected()) {
        return WiFi.localIP().toString();
    }
    return "";
}

// 获取网络信息
NetworkInfo WiFiManager::getNetworkInfo() {
    NetworkInfo info;
    
    if (isConnected()) {
        info.ip = WiFi.localIP().toString();
        info.subnet = WiFi.subnetMask().toString();
        info.gateway = WiFi.gatewayIP().toString();
        info.dns1 = WiFi.dnsIP().toString();
        info.dns2 = WiFi.dnsIP(1).toString();
        info.rssi = WiFi.RSSI();
        info.mac = WiFi.macAddress();
        info.hostname = WiFi.getHostname();
    }
    
    return info;
}

// 获取WiFi信号强度
int WiFiManager::getRSSI() {
    if (isConnected()) {
        return WiFi.RSSI();
    }
    return 0;
}

// 获取MAC地址
String WiFiManager::getMacAddress() {
    return WiFi.macAddress();
}

// 设置主机名
bool WiFiManager::setHostname(const String& hostname) {
    return WiFi.setHostname(hostname.c_str());
}

// 扫描可用WiFi网络
int WiFiManager::scanNetworks() {
    log("开始扫描WiFi网络...");
    int networks = WiFi.scanNetworks();
    log("发现 " + String(networks) + " 个网络");
    return networks;
}

// 扫描网络并验证目标SSID是否存在
bool WiFiManager::scanAndVerifySSID() {
    try {
        log("扫描可用的WiFi网络...");
        int networkCount = WiFi.scanNetworks();
        log("发现 " + String(networkCount) + " 个网络:");
        
        bool ssidFound = false;
        for (int i = 0; i < networkCount; i++) {
            String scannedSSID = WiFi.SSID(i);
            int rssi = WiFi.RSSI(i);
            log("  [" + String(i) + "] " + scannedSSID + " (信号强度: " + String(rssi) + " dBm)");
            
            if (scannedSSID == ssid) {
                ssidFound = true;
                log("  ✓ 找到目标网络: " + ssid + " (信号强度: " + String(rssi) + " dBm)");
            }
        }
        
        if (!ssidFound) {
            log("❌ 错误: 未找到指定的网络 '" + ssid + "'");
            log("请检查:");
            log("  1. SSID名称是否正确（区分大小写）");
            log("  2. 设备是否在WiFi覆盖范围内");
            log("  3. 网络是否为2.4GHz频段");
            log("  4. 路由器是否启用SSID广播");
            return false;
        }
        
        return true;
        
    } catch (...) {
        log("网络扫描失败: 未知异常");
        return false;
    }
}

// 获取扫描到的网络信息
String WiFiManager::getScannedSSID(int index) {
    if (index >= 0 && index < WiFi.scanComplete()) {
        return WiFi.SSID(index);
    }
    return "";
}

// 清理资源并释放内存
void WiFiManager::cleanup() {
    try {
        if (WiFi.status() == WL_CONNECTED) {
            WiFi.disconnect(true);
        }
        WiFi.mode(WIFI_OFF);
        connected = false;
        logCallback = nullptr;
        log("WiFi 资源已清理");
    } catch (...) {
        Serial.println("WiFi 清理异常");
    }
}

// ==================== NTPTimeSync 实现 ====================

// 常用NTP服务器列表
const char* NTPTimeSync::NTP_SERVERS[] = {
    "ntp.aliyun.com",       // 阿里云
    "ntp.ntsc.ac.cn",       // 中国国家授时中心
    "ntp1.aliyun.com",      // 阿里云备用
    "pool.ntp.org",         // 国际NTP池
    "time.nist.gov",         // NIST时间服务器
    "time.asia.apple.com",
    "time.apple.com"
};

const int NTPTimeSync::NTP_SERVER_COUNT = sizeof(NTP_SERVERS) / sizeof(NTP_SERVERS[0]);

// 构造函数
NTPTimeSync::NTPTimeSync(int timezoneOffsetHours)
    : timezoneOffset(timezoneOffsetHours * 3600), logCallback(nullptr) {
}

// 析构函数
NTPTimeSync::~NTPTimeSync() {
    // ESP32 Arduino框架会自动管理SNTP资源
    // 无需手动停止
}

// 设置日志回调函数
void NTPTimeSync::setLogCallback(LogCallback callback) {
    logCallback = callback;
}

// 内部日志方法
void NTPTimeSync::log(const String& message) {
    if (logCallback) {
        logCallback(message);
    } else {
        Serial.println(message);
    }
}

// 从NTP服务器同步时间
bool NTPTimeSync::sync(const char* ntpServer, int retryCount) {
    try {
        // 确定要使用的NTP服务器列表
        const char* servers[NTP_SERVER_COUNT + 1];
        int serverCount = 0;
        
        if (ntpServer) {
            servers[serverCount++] = ntpServer;
        }
        
        // 添加默认服务器
        for (int i = 0; i < NTP_SERVER_COUNT; i++) {
            servers[serverCount++] = NTP_SERVERS[i];
        }
        
        // 尝试每个服务器
        for (int s = 0; s < serverCount; s++) {
            for (int attempt = 0; attempt < retryCount; attempt++) {
                try {
                    log("正在从 " + String(servers[s]) + " 同步时间... (尝试 " + 
                        String(attempt + 1) + "/" + String(retryCount) + ")");
                    
                    // 配置时区和NTP服务器
                    configTime(timezoneOffset, 0, servers[s]);
                    
                    // 等待时间同步
                    int syncTimeout = 20000;  // 20秒超时
                    unsigned long startTime = millis();
                    
                    while (!isTimeSynced() && (millis() - startTime < syncTimeout)) {
                        delay(100);
                    }
                    
                    if (isTimeSynced()) {
                        // 同步成功
                        String currentTime = getISO8601Time();
                        log("时间同步成功: " + currentTime);
                        return true;
                    } else {
                        log("同步超时");
                    }
                    
                } catch (...) {
                    log("同步异常");
                    if (attempt < retryCount - 1) {
                        delay(2000);  // 重试前等待
                    }
                    continue;
                }
            }
        }
        
        log("所有 NTP 服务器同步失败");
        return false;
        
    } catch (...) {
        log("时间同步异常");
        return false;
    }
}

// 设置时区偏移量
void NTPTimeSync::setTimezoneOffset(int hours) {
    timezoneOffset = hours * 3600;
}

// 获取当前时间戳
time_t NTPTimeSync::getTimestamp() {
    time_t now;
    time(&now);
    return now;
}

// 获取ISO 8601格式的当前时间字符串
String NTPTimeSync::getISO8601Time() {
    time_t now = getTimestamp();
    struct tm* timeinfo = localtime(&now);
    
    char buffer[32];
    snprintf(buffer, sizeof(buffer), "%04d-%02d-%02dT%02d:%02d:%02d",
             timeinfo->tm_year + 1900,
             timeinfo->tm_mon + 1,
             timeinfo->tm_mday,
             timeinfo->tm_hour,
             timeinfo->tm_min,
             timeinfo->tm_sec);
    
    return String(buffer);
}

// 获取带时区偏移的ISO 8601格式时间
String NTPTimeSync::getISO8601TimeWithTimezone(int timezoneOffsetHours) {
    time_t now = getTimestamp();
    struct tm* timeinfo = localtime(&now);
    
    // 计算时区偏移的小时和分钟
    int offsetHours = timezoneOffsetHours;
    int offsetMinutes = 0;
    char offsetSign = (offsetHours >= 0) ? '+' : '-';
    offsetHours = abs(offsetHours);
    
    char buffer[64];
    snprintf(buffer, sizeof(buffer), "%04d-%02d-%02dT%02d:%02d:%02d%c%02d:%02d",
             timeinfo->tm_year + 1900,
             timeinfo->tm_mon + 1,
             timeinfo->tm_mday,
             timeinfo->tm_hour,
             timeinfo->tm_min,
             timeinfo->tm_sec,
             offsetSign,
             offsetHours,
             offsetMinutes);
    
    return String(buffer);
}

// 获取UTC时间的ISO 8601格式
String NTPTimeSync::getISO8601TimeUTC() {
    time_t now = getTimestamp();
    struct tm* timeinfo = gmtime(&now);
    
    char buffer[32];
    snprintf(buffer, sizeof(buffer), "%04d-%02d-%02dT%02d:%02d:%02dZ",
             timeinfo->tm_year + 1900,
             timeinfo->tm_mon + 1,
             timeinfo->tm_mday,
             timeinfo->tm_hour,
             timeinfo->tm_min,
             timeinfo->tm_sec);
    
    return String(buffer);
}

// 格式化当前时间
String NTPTimeSync::formatTime(const String& formatStr) {
    time_t now = getTimestamp();
    struct tm* timeinfo = localtime(&now);
    
    char buffer[128];
    strftime(buffer, sizeof(buffer), formatStr.c_str(), timeinfo);
    
    return String(buffer);
}

// 检查时间是否已同步
bool NTPTimeSync::isTimeSynced() {
    time_t now = getTimestamp();
    return (now > 1000000000);  // 时间戳大于2001年表示已同步
}

// 获取系统运行时间（毫秒）
unsigned long NTPTimeSync::getUptime() {
    return millis();
}

// 延迟到指定时间（非阻塞检查）
bool NTPTimeSync::delayUntil(time_t targetTime) {
    return getTimestamp() >= targetTime;
}

// ==================== 便捷函数实现 ====================

// 快速连接WiFi（便捷函数）
WiFiManager* quickConnectWifi(const String& ssid, const String& password, int timeout) {
    WiFiManager* wifi = new WiFiManager(ssid, password);
    if (wifi->connect(timeout)) {
        return wifi;
    }
    delete wifi;
    return nullptr;
}

// 快速同步时间（便捷函数）
bool quickSyncTime(int timezoneOffsetHours, const char* ntpServer) {
    NTPTimeSync ntp(timezoneOffsetHours);
    return ntp.sync(ntpServer);
}

// 检查网络连接状态
bool isNetworkConnected() {
    return (WiFi.status() == WL_CONNECTED);
}

// 获取网络状态字符串
String getNetworkStatus() {
    switch (WiFi.status()) {
        case WL_IDLE_STATUS:
            return "空闲状态";
        case WL_NO_SSID_AVAIL:
            return "无可用网络";
        case WL_SCAN_COMPLETED:
            return "扫描完成";
        case WL_CONNECTED:
            return "已连接 - IP: " + WiFi.localIP().toString();
        case WL_CONNECT_FAILED:
            return "连接失败";
        case WL_CONNECTION_LOST:
            return "连接丢失";
        case WL_DISCONNECTED:
            return "已断开";
        case WL_NO_SHIELD:
            return "无WiFi模块";
        default:
            return "未知状态: " + String(WiFi.status());
    }
}

// 等待网络连接
bool waitForNetwork(int timeout) {
    unsigned long startTime = millis();
    
    while (!isNetworkConnected() && (millis() - startTime < timeout * 1000UL)) {
        delay(100);
    }
    
    return isNetworkConnected();
}